execution/order_by.go

```diff
@@ -18,132 +15,217 @@ const (
+		return nil, nil, errors.Wrap(err, "couldn't get trigger from trigger prototype")
+	}
+
+	node.key = make([]Expression, 0)
+	if len(node.eventTimeField) > 0 {
+		node.key = append(node.key, node.expressions[0])
+	} else {
+		node.key = append(node.key, NewConstantValue(octosql.MakeString("key")))
 	}
 
-	return orderedStream, NewExecutionOutput(NewZeroWatermarkGenerator(), execOutput.NextShuffles, execOutput.TasksToRun), nil
+	orderBy := &OrderByStream{
+		variables:       variables,
+		expressions: node.expressions,
+		directions:  node.directions,
+	}
+
+	processFunc := &ProcessByKey{
+		eventTimeField:  node.eventTimeField,
+		trigger:         trigger,
+		keyExpressions:  [][]Expression{node.key},
+		processFunction: orderBy,
+		variables:       variables,
+	}
+
+	orderByPullEngine := NewPullEngine(processFunc, node.storage, []RecordStream{source}, streamID, execOutput.WatermarkSource, true, ctx)
+
+	return orderByPullEngine, // groupByPullEngine now indicates new watermark source
+		NewExecutionOutput(
+			orderByPullEngine,
+			execOutput.NextShuffles,
+			append(execOutput.TasksToRun, func() error { orderByPullEngine.Run(); return nil }),
+		),
+		nil
 }
 
-func createOrderedStream(ctx context.Context, expressions []Expression, directions []OrderDirection, variables octosql.Variables, sourceStream RecordStream) (stream RecordStream, outErr error) {
-	records := make([]*Record, 0)
+type OrderByStream struct {
+	expressions []Expression
+	directions  []OrderDirection
+	variables       octosql.Variables
+}
 
-	for {
-		rec, err := sourceStream.Next(ctx)
-		if err == ErrEndOfStream {
-			break
-		} else if err != nil {
-			return nil, errors.Wrap(err, "couldn't get all records")
-		}
+var recordValuePrefix = []byte("$record_value$")
+//todo record count prefix is from orderby
 
-		records = append(records, rec)
+func (ob *OrderByStream) AddRecord(ctx context.Context, tx storage.StateTransaction, inputIndex int, key octosql.Value, record *Record) error {
+	if inputIndex > 0 {
+		panic("only one input stream allowed for order by")
 	}
 
-	defer func() {
-		if err := recover(); err != nil {
-			stream = nil
-			outErr = errors.Wrap(err.(error), "couldn't sort records")
+	keyPrefix := append(append([]byte("$"), key.MonotonicMarshal()...), '$')
+	txByKey := tx.WithPrefix(keyPrefix)
+
+	var recordValueMap *storage.Map
+	var recordCountMap *storage.Map
+
+
+	recordValueMap = storage.NewMap(txByKey.WithPrefix(recordValuePrefix))
+	recordCountMap = storage.NewMap(txByKey.WithPrefix(recordCountPrefix))
+
+	mapping := make(map[string]octosql.Value, len(record.Fields()))
+	for _, field := range record.Fields() {
+		mapping[field.Name.String()] = record.Value(field.Name)
+	}
+	variables := record.AsVariables()
+
+
+	recordPrefix := []byte("$")
+
+	for i := range ob.expressions {
+		expressionValue, err := ob.expressions[i].ExpressionValue(ctx, variables)
+		if err != nil {
+			return errors.Wrap(err, "couldn't evaluate expression")
```

Comment:
You could use `Wrapf` and include the index of the expression like "couldn't evalute expression with index %d", i